### ISSUE TITLE
Disable scripting for documents without browsing contexts

### ISSUE THREAD

**Issue by jdm** [servo/servo#10940]
Document contains a `scripting_enabled` flag which is currently always true. We should instead set it to false if the document has no browsing context when it's created.

**Specification**: https://html.spec.whatwg.org/multipage/webappapis.html#concept-n-script
**Code**: `components/script/dom/document.rs`
**Test**: let's [write a test](https://github.com/servo/servo/blob/master/tests/wpt/README.md#writing-new-tests) in `tests/wpt/web-platform-tests/html/semantics/scripting-1/the-script-element/` that uses [createDocument()](https://dom.spec.whatwg.org/#dom-domimplementation-createdocument) and verifies that `<script>` elements inserted into the document don't execute.

**Comment by bd339** [comment]
I would like to work on this.

**Comment by jdm** [comment]
@bd339 Go for it! Please ask questions about anything that's unclear.

**Comment by bd339** [comment]
For writing the test... I thought about inserting a `<script>` element into a document returned by `createDocument`, that will do something equivalent to `assert_equals(true, false)`. Would this work?

**Comment by jdm** [comment]
We would probably need something like `new_doc.script_executed = t.unreached_func("shouldn't be called");` and have the script invoke document.scrip_executed().

**Comment by bd339** [comment]
I ran this command to create the new test:
`./mach create-wpt tests/wpt/web-platform-tests/html/semantics/scripting-1/the-script-element/scripting-enabled.html`
But when I try to run the test with this command:
`./mach test-wpt tests/wpt/web-platform-tests/html/semantics/scripting-1/the-script-element/scripting-enabled.html`
... it reports that it has finished running 0 tests. Why is this?

**Comment by jdm** [comment]
Did you interrupt the create-wpt command while it was running? You'll need to run `--manifest-update` to the test-wpt command next time to make it work.

**Comment by bd339** [comment]
This is the command I ran along with the ouptut:
`./mach create-wpt tests/wpt/web-platform-tests/html/semantics/scripting-1/the-script-element/scripting-enabled.html`

> Running virtualenv with interpreter /usr/bin/python2.7
New python executable in /home/benjamin/bd339/servo/python/_virtualenv/bin/python2.7
Also creating executable in /home/benjamin/bd339/servo/python/_virtualenv/bin/python
Installing setuptools, pip...done.
looking for rustc at /home/benjamin/bd339/servo/.servo/rust/2016-04-24/rustc-nightly-x86_64-unknown-linux-gnu/rustc/bin/rustc
Downloading Rust compiler...
Downloading Rust compiler: 100.0%
Extracting Rust compiler...
Rust compiler ready.
Downloading Host rust library for target x86_64-unknown-linux-gnu...
Downloading Host rust library for target x86_64-unknown-linux-gnu: 100.0%
Extracting Rust stdlib for target x86_64-unknown-linux-gnu...
Rust x86_64-unknown-linux-gnu libs ready.
Downloading Cargo nightly...
Downloading Cargo nightly: 100.0%
Extracting Cargo nightly...
Cargo ready.
Binary path /home/benjamin/bd339/servo/target/debug/servo does not exist

**Comment by bd339** [comment]
I'm facing the difficulty that even when `scripting_enabled` is always true, I can't get the test to fail. This is the test body.

var doc = document.implementation.createDocument ('http://www.w3.org/1999/xhtml', 'html');
doc.fail_test = t.unreached_func('should not have been called');

var script = document.createElementNS('http://www.w3.org/1999/xhtml', 'script');
script.text = 'document.fail_test();';

doc.documentElement.appendChild(script);

Note that the test does fail if I explicitly invoke `doc.fail_test();`. Otherwise it will just time out.

**Comment by jdm** [comment]
With respect to the create-wpt command, have you completed a debug build of servo?

**Comment by bd339** [comment]
That was just before I did a debug build of my fork, so I guess that part is expected.

**Comment by jdm** [comment]
Ah, I figured out why the script doesn't execute correctly  - HTMLScriptElement::execute obtains the window for the document and calls evaluate_script_on_global_with_result, so the object for `document` is not the same as the newly created one. This triggers for me:
```html
<script>
var d =  document.implementation.createHTMLDocument();
d.fail_test = function() { console.log("hi") };
var s = d.createElement('script');
s.textContent = 'd.fail_test()';
d.documentElement.appendChild(s);
</script>
```

**Comment by jdm** [comment]
It might make more sense to create a `fail_test` function on the global window scope and have the script content be `fail_test()`, in that case.

**Comment by KiChjang** [comment]
Fixed in #10946.
